### PR TITLE
Add pylint

### PR DESCRIPTION
**Pull request by alexfikl**
Most of the setup for pylint was copied from pytential.

There's a CI running on gitlab over at https://gitlab.tiker.net/fikl2/meshmode/-/pipelines/139540

### CODE REVIEW

File: meshmode/mesh/__init__.py [hunk 1]
```diff
@@ -409,7 +403,7 @@ class NodalAdjacency(Record):
 
     def __eq__(self, other):
         return (
-                type(self) == type(other)
+                isinstance(self, type(other))
```
Comment (inducer) [inline meshmode/mesh/__init__.py:406]: Nope

File: meshmode/mesh/__init__.py [hunk 2]
```diff
@@ -481,7 +475,7 @@ class FacialAdjacencyGroup(Record):
...
```
Comment (inducer) [inline meshmode/mesh/__init__.py:478]: Nope

File: meshmode/mesh/__init__.py [hunk 3]
```diff
@@ -866,10 +860,12 @@ def nelements(self):
     @property
     def nodal_adjacency(self):
         from meshmode import DataUnavailable
+
+        # pylint: disable=E0203
```
Comment (inducer) [inline meshmode/mesh/__init__.py:864]: Symbolic name

File: meshmode/mesh/__init__.py [hunk 4]
```diff
@@ -866,10 +860,12 @@ def nelements(self):
...
         if self._nodal_adjacency is False:
             raise DataUnavailable("nodal_adjacency")
 
-        elif self._nodal_adjacency is None:
+        if self._nodal_adjacency is None:
```
Comment (inducer) [inline meshmode/mesh/__init__.py:868]: Why is that an improvement?

File: meshmode/mesh/__init__.py [hunk 5]
```diff
@@ -888,10 +884,12 @@ def nodal_adjacency_init_arg(self):
     @property
     def facial_adjacency_groups(self):
         from meshmode import DataUnavailable
+
+        # pylint: disable=E0203
```
Comment (inducer) [inline meshmode/mesh/__init__.py:888]: Symbolic name

File: meshmode/mesh/__init__.py [hunk 6]
```diff
@@ -910,7 +908,7 @@ def boundary_tag_bit(self, boundary_tag):
 
     def __eq__(self, other):
         return (
-                type(self) == type(other)
+                isinstance(self, type(other))
```
Comment (inducer) [inline meshmode/mesh/__init__.py:911]: Nope

File: meshmode/mesh/__init__.py [hunk 7]
```diff
@@ -888,10 +884,12 @@ def nodal_adjacency_init_arg(self):
     @property
     def facial_adjacency_groups(self):
         from meshmode import DataUnavailable
+
+        # pylint: disable=E0203
         if self._facial_adjacency_groups is False:
             raise DataUnavailable("facial_adjacency_groups")
 
-        elif self._facial_adjacency_groups is None:
+        if self._facial_adjacency_groups is None:
```
Comment (inducer) [inline meshmode/mesh/__init__.py:892]: ?

**Review by inducer — COMMENTED** [review]
Thanks! I like *most* of these changes. Plus an additional linter, yay! But some were puzzling to me. Could you take a look?

File: meshmode/mesh/__init__.py [hunk 8]
```diff
@@ -277,7 +271,7 @@ def nfaces(self):
 
     def __eq__(self, other):
         return (
-                type(self) == type(other)
+                isinstance(self, type(other))
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/__init__.py]: Reverted in a16b6cd.

They came from [C0123 unidiomatic-typecheck](https://pycodequ.al/docs/pylint-messages/c0123-unidiomatic-typecheck.html).

File: meshmode/mesh/processing.py [hunk 9]
```diff
@@ -122,8 +123,6 @@ def _filter_mesh_groups(groups, selected_elements, vertex_id_dtype):
     for i_new_grp, i_old_grp in enumerate(new_group_to_old_group):
         group_to_new_group[i_old_grp] = i_new_grp
 
-    del grp
-
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/processing.py]: Reverted in 5c62146.

File: test/test_meshmode.py [hunk 10]
```diff
@@ -1604,7 +1590,7 @@ def test_mesh_multiple_groups(actx_factory, ambient_dim, visualize=False):
             axis=1).astype(np.int64)
     mesh = split_mesh_groups(mesh, element_flags)
 
-    assert len(mesh.groups) == 2
+    assert len(mesh.groups) == 2            # pylint: disable=E1101
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply test/test_meshmode.py]: Fixed all of them in 856038e

File: meshmode/mesh/__init__.py (hunk 4 above)
Comment (alexfikl) [reply meshmode/mesh/__init__.py]: Reverted in f8b5437.

This came from [R1720 no-else-raise](https://pycodequ.al/docs/pylint-messages/r1720-no-else-raise.html).

**Comment by alexfikl** [comment]
> Thanks! I like _most_ of these changes. Plus an additional linter, yay! But some were puzzling to me. Could you take a look?

I ran this first with a "raw" pylint, which complained about a LOT of things, before using the setup from pytential and ended up "fixing" a bit too much. Sorry for the noise!

File: meshmode/mesh/refinement/__init__.py [hunk 11]
```diff
@@ -20,6 +20,7 @@
 THE SOFTWARE.
 """
 
+# pylint: disable-all
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/refinement/__init__.py]: Actually re-enabled this in ce2ddb2.

It had a huge amount of warnings in the Convention/Recommendation categories, but those are disabled by default with the setup from `ci-support`.

**Comment by alexfikl** [comment]
Latest commits running on gitlab
https://gitlab.tiker.net/fikl2/meshmode/-/pipelines/139675

**Comment by inducer** [comment]
Thanks!

**Pull request merged**